kq_epic_post.md

```diff
@@ -0,0 +1,160 @@
+    AND `issues`.`pull_request_id` = `pull_requests`.`id`
+    AND `issues`.`repository_id` = `pull_requests`.`repository_id`
+    AND `issues_labels`.`issue_id` = `issues`.`id`
+    AND `labels`.`id` = `issues_labels`.`label_id`
+    AND `labels`.`repository_id` = `issues`.`repository_id`
+  ORDER BY
+    `pull_requests`.`id` DESC LIMIT ?
+```
+
+Simply swapping `issues.state` for `pull_request.status` wasn't enough for these queries; they were still timing out. Thus we explored other optimizations, including breaking out subqueries, forcing or ignoring an index, and reorganizing the query structure, none of which showed improvements[^1]. We even considered storing label ids in a new column on `issues` to eliminate the extra joins, but that presented its own big lift that seemed far outside the scope of this epic.
+
+So, like the dwarves in the mines of Moria, we dug deeper to determine _who_ was hitting these queries. Over a period of 30 days, we found that between the two queries, all calls were coming from only 4 users within 4 repos. More interestingly, all 4 users were GitHub apps mimicking merge queue functionality. One in particular was causing around 10k events routinely every 20 minutes. With this new information, our minds turned towards rate limiting. Unfortunately, rate limits are set based on individual app installations and cannot be applied by query fingerprint, so they were too blunt to deploy here.
+
+Once we'd exhausted our best efforts, we ultimately decided to forgo this particular quest lest we awaken the Balrog. Due to the very small group of affected users, we determined that the estimated effort to continue was not worth the potential impact. Instead, there is an open discussion about updating our scorecard tooling to exempt these queries from our scorecards moving forward.
+
+#### Obstacle 3: The road goes ever on
+To populate our new `pull_requests.status`column, we needed to backfill with the data from issues.state. `pull_requests` is a _big_ table, so we knew we were in for a hefty process. What we didn't expect is that the transition would take 604 hours of run time across a period of 33 days, including dry runs. Within that, we had to maintain vigilance as we encountered unexplained failures that would interrupt the process, requiring us to manually restart. Finally, when we thought our toil had borne its reward, we were faced with some confusing data.
+
+While the logging indicated everything had run as expected, Kusto, datadot, and analytics hosts suggested the backfill had missed some pull requests[^2]. This was potentially an enormous setback. To find a reliable source of truth, we reached out to several teams and finally learned that what we needed was a new full snapshot of the table. With that done, we were able to determine that the backfill was mostly successful and the remaining group of pull requests missing `status` data were from the period after dual writing was introduced.
```

Comment:
This is actually a really useful learning and it might be worth linking to references of the differences in the snapshot types and what was wrong with the snapshot you had!

Reply:
I discovered we needed this from a response to asking in the data-engineering Slack channel, but I couldn't find any docs where it's mentioned.